### DISCUSSION TITLE
How to properly build NextJS application?

### DISCUSSION THREAD

**Discussion by iAmGiganticFlower** [vercel/next.js#64724] [Help] [answered]
### Summary

My `Dockerfile` usually has this commands to build the NextJS application:
```Dockerfile
npm install
npm run build
```

And almost all the time I was ok with that

But now I realised, that there's many `devDependencies` which I would like to exclude from my build because I don't need the at all when I build my application in production (because they increase `npm install` time of my application)

So I added env:
```Dockerfile
NODE_ENV="production"
```

And now my `npm install` installs only `dependencies` without `devDependencies` and I am glad for that

But after that my `npm run build` (which is `next build`) is also got broken, because it runs `ts lint` during building and I get errors kinda like that:
```
Cannot resolve @types/..., try to install them using @types/...
```
because my `devDependencies` are missing

What should I do in this case?

I don't want to move `devDependencies` to the `dependencies`, I also don't want to remove `NODE_ENV="production"` from my `Dockerfile`

What's the most common practice for my case?

Thank you

**Comment by GabenGar** [comment, marked as answer]
Build in a separate image and copy artifacts into production one. You aren't going to win the fight against the framework on its own turf.
